### ISSUE TITLE
[Feature Request] Option to delete file revisions

### ISSUE THREAD

**Issue by zDEFz** [zadam/trilium#1528]
Having 1GiB data in my trilium-Notes. Trilium bloats to 8GiB.

Need a way to clean out...

**Comment by zadam** [comment]
It's possible to delete revisions for a particular note:

![](https://user-images.githubusercontent.com/617641/104169405-cf066200-53ff-11eb-9f10-33317a078aa1.png)

**Comment by zDEFz** [comment]
@zadam thats not feasible.
I want to delete all revisions of all note at once cause I don't know why my database is so bloated.

I also can't just export all my files in raw format, because Trilium note crashes then.
I had to import my files in small zip archive chunks, cause as soon as it got big = crash.

If I could export all my files out, then create a new database and re-import them, the database would be 8 times smaller, I promise.

**Comment by zDEFz** [comment]
---
Actually, are you aware of those problems? Would it help if we had a call?
Do you value my input?

I mean, nothing really is working as it should.

When you cant import files properly, sync also fails.
And no, my fails aren't exceeding the limits.
When I setup sync to a target server within trilium, it will get to the point where trilium desktop app crashes on startup.

Both Windows and Mac. Cause a single file fails at sync.
All you can do is to create some temporary files and really really quickly set sync to empty values again...

**Comment by zadam** [comment]
Frankly I think most of your problems come from the fact that you use trilium in a way it was not envisioned for. It was never intended to be used for big file storage. As a consequence these use cases are not properly tested and there are these problems.

**Comment by zDEFz** [comment]
I argue and say "think big" [...]

There is a reason I want to use trilium for that use-case. It has so far the best structure for me.
I use it as an Accessibility aid, so to say. That's why I personally think, it should be more robust.
I haven't yet understood how I can help develop on this application.

I'm doubting my ability to programmatically enhance it's capability on my own.

If you know something better suited for the job, please let me know. I personally don't know any application that is fitting this kind of "simplicity in complexity" and ease of use.

**Comment by wwwald** [comment]
@zadam: Would there be way to do this directly in the SQLite database? 

I could imagine a query or script involving `note_revisions` and `note_revision_contents` that may be able to wipe out lots of revisions at once?  
Or does that risk introducing inconsistencies in the database?

I'm willing to _try_ to help out, but I actually agree with @zadam - this is a quite specific niche use case that needs specific optimizations in the code. I'm not too familiar with the codebase myself to do reasonable suggestions yet...

**Comment by zDEFz** [comment]
@zadam I wonder if it would help would I send you the "broken" database in anon form.
I mean, as soon as you try to sync from it you are stuck in a crash on startup loop.
Or alternatively, ask me to deleted private parts in it.

**Comment by zadam** [comment]
@wwwald yes, you could clean them out manually using SQL like that. After that `VACUUM` is needed to actually shrink the database file.

@zDEFz in this case anonymized document is not very useful since all those big blobs will be stripped. So unanonymized database would be help. I could fix at least the low hanging fruit.

BTW, for 0.46 I did already some big file optimizations and currently I'm working on a sync protocol change which will mitigate big file syncing issues caused by proxy limits by chunking up the big requests into smaller ones. But the fact will remain that Trilium is not a system for file storage. I'm wondering how to make that clear - I don't want to set some specific limits, but perhaps I'll hang a clearly visible warning for too large files (e.g. over 50 MBs).

**Comment by zDEFz** [comment]
Ver:
0.45.7
@zadam If that would be the problem it would be easier to fix.

No file within this node is over 50 MBs, but it will fail to export or sync!
I have created the document.db completely new so it's unlikely to be the problem.

You will get: Export failed with following error: 'Cannot read property 'length' of null'. More details might be in the logs.
[trilium-2021-01-13.log](https://github.com/zadam/trilium/files/5808339/trilium-2021-01-13.log)

The import will be succesful, the export unsuccesful and the sync leading to a crash loop on startup.

Thanks for Mutopia to provide their files publicly in a FTP...

https://mega.nz/file/d483DC6I#xHMoqCufuSL7AE-_wmxt7N7IutXPXtQRv0O_veLgk4Y

**Comment by zDEFz** [comment]
-add : I agree hinting the user to not add too huge files, but my problem is not my fault, I think.

**Comment by wadbr** [comment]
Dude just for your information but relying on a note application to keep your data "safe" is never a good idea. You need actual backups and things designed to keep data safe and uncorruptuable.

**Comment by zDEFz** [comment]
> Dude just for your information but relying on a note application to keep your files/data "safe" is never a good idea. You need actual backups and things designed to keep non-text files safe and uncorruptuable.

Thanks for your concern. But please don't go off-topic. I have my backups in place.

**Comment by zDEFz** [comment]
@zadam temporary solution was to duplicate a subtree and then remove the original, clicking remove all clones also.
Saved 7GB...
